File: f77/driver/rmevent.f
C
      Subroutine Remove_Event
C
C     This subroutine removes the first event in the future event
C     queue. All subsequent events are then advanced in the queue.
C
C     Inputs:
C        Events            - Event queue [Integer Array]
C        Number_Of_Events  - Current number of events in the event queue [Integer]
C        Time_Of_Events    - Time to execute corresponding event in event queue [Real Array]
C
C     Outputs:
C        Events            - Event queue [Integer Array]
C        Number_Of_Events  - Current number of events in the event queue [Integer]
C        Time_Of_Events    - Time to execute corresponding event in event queue [Real Array]
C
C     Internal variables and constants:
C        I                 - Loop Index [Integer]
C
      Include 'global.inc'
      Include 'sysvars.inc'
      Include 'event.inc'
C
C     Exceptions to default type
C
      Integer I
C
C     Remove the first event and shift down future events list.
C     Check for valid Number_Of_Events
C
      If (Number_Of_Events .gt. 0) Then
         I = 2
         Do While (I .le. Number_Of_Events)
            Time_Of_Events(I - 1) = Time_Of_Events(I)
            Events (I - 1) = Events(I)
            I = I + 1
         End do
         Time_Of_Events(Number_of_Events) = 99999.0E0
         Events(Number_Of_Events)         = 0
         Number_Of_Events                 = Number_Of_Events - 1
C
C     Inform user if error condition exists
C
      Else
         Write (*, '(A)')
     &        'WARNING : event removal requested from empty queue.'
      End If
C
      Return
      End

File: compcert/driver/rmevent.c
#include "../driver/global.h"
#include "../driver/sysvars.h"
#include "../driver/event.h"
void remove_event() {
  // This subroutine removes the first event in the future event queue. All
  // subsequent events are then advanced in the queue.
  //
  // Inputs:
  //     Events            - Event queue [Integer Array]
  //     Number_Of_Events  - Current number of events in the event queue [Integer]
  //     Time_Of_Events    - Time to execute corresponding event in event queue [Real Array]
  //
  // Outputs:
  //     Events            - Event queue [Integer Array]
  //     Number_Of_Events  - Current number of events in the event queue [Integer]
  //     Time_Of_Events    - Time to execute corresponding event in event queue [Real Array]

  // Remove the first event and shift down future events list.
  // Check for valid Number_Of_Events
  int i;
  if (number_of_events > 0) {
    i = 1;
    while (i <= number_of_events) {
      time_of_events[i-1] = time_of_events[i];
      events[i-1] = events[i];
      i += 1;
    }
    time_of_events[number_of_events] = 1.0e10;
    events[number_of_events] = 0;
    number_of_events -= 1;
    // Inform user if error condition exists
  }
  else
    printf(" WARNING : event removal requested from empty queue.\n");
}
